bd-resilient-kv/src/kv_store.rs

```diff
@@ -0,0 +1,246 @@
+
+    // Try to open with existing data first
+    MemMappedKVJournal::from_file(path, high_water_mark_ratio, callback).or_else(|_| {
+      // Data is corrupt or unreadable, create fresh
+      MemMappedKVJournal::new(path, target_size, high_water_mark_ratio, callback)
+    })
+  }
+
+  /// Get a value by key.
+  ///
+  /// This operation is O(1) as it reads from the in-memory cache.
+  #[must_use]
+  pub fn get(&self, key: &str) -> Option<Value> {
+    self.cached_map.get(key).cloned()
+  }
+
+  /// Insert a value for a key, returning the previous value if it existed.
+  ///
+  /// Note: Inserting `Value::Null` is equivalent to removing the key.
+  ///
+  /// # Errors
+  /// Returns an error if the value cannot be written to the journal.
+  pub fn insert(&mut self, key: String, value: Value) -> anyhow::Result<Option<Value>> {
+    let old_value = self.get(&key);
+    if matches!(value, Value::Null) {
+      // Inserting null is equivalent to deletion
+      if old_value.is_some() {
+        self.journal.delete(&key)?;
+        self.cached_map.remove(&key);
+      }
+    } else {
+      self.journal.set(&key, &value)?;
+      self.cached_map.insert(key, value);
+    }
+    Ok(old_value)
+  }
+
+  /// Remove a key and return its value if it existed.
+  ///
+  /// # Errors
+  /// Returns an error if the deletion cannot be written to the journal.
+  pub fn remove(&mut self, key: &str) -> anyhow::Result<Option<Value>> {
+    let old_value = self.get(key);
+    if old_value.is_some() {
+      self.journal.delete(key)?;
+      self.cached_map.remove(key);
+    }
+    Ok(old_value)
+  }
+
+  /// Check if the store contains a key.
+  ///
+  /// This operation is O(1) as it reads from the in-memory cache.
+  #[must_use]
+  pub fn contains_key(&self, key: &str) -> bool {
+    self.cached_map.contains_key(key)
+  }
+
+  /// Get the number of key-value pairs in the store.
+  ///
+  /// This operation is O(1) as it reads from the in-memory cache.
+  #[must_use]
+  pub fn len(&self) -> usize {
+    self.cached_map.len()
+  }
+
+  /// Check if the store is empty.
+  ///
+  /// This operation is O(1) as it reads from the in-memory cache.
+  #[must_use]
+  pub fn is_empty(&self) -> bool {
+    self.len() == 0
+  }
+
+  /// Clear all key-value pairs from the store.
+  ///
+  /// # Errors
+  /// Returns an error if the clearing operation fails.
+  pub fn clear(&mut self) -> anyhow::Result<()> {
+    self.journal.clear()?;
+    self.cached_map.clear();
+    Ok(())
+  }
+
+  /// Get all keys in the store.
+  ///
+  /// This operation returns an iterator over the keys without allocation.
+  #[must_use]
+  pub fn keys(&self) -> std::collections::hash_map::Keys<'_, String, Value> {
+    self.cached_map.keys()
+  }
+
+  /// Get all values in the store.
+  ///
+  /// This operation returns an iterator over the values without allocation.
+  #[must_use]
+  pub fn values(&self) -> std::collections::hash_map::Values<'_, String, Value> {
+    self.cached_map.values()
+  }
```

Comment:
Is any of this used? I would just remove all of this extra stuff if you provide ref access to the hash map, as anything can be done from that.